app/models/filtering_test.rb

```diff
@@ -1,3 +1,118 @@
 class FilteringTest < ProductTest
+  field :options, type: Hash
   accepts_nested_attributes_for :tasks
+
+  after_create :create_tasks
+
+  def create_tasks
+    tasks.build({ product_test: self }, Cat1FilterTask).save
+    tasks.build({ product_test: self }, Cat3FilterTask).save
+    save
+  end
+
+  def cat1_task
+    cat1_tasks = tasks.select { |task| task._type == 'Cat1FilterTask' }
+    if cat1_tasks.empty?
+      false
+    else
+      cat1_tasks.first
+    end
+  end
+
+  def cat3_task
+    cat3_tasks = tasks.select { |task| task._type == 'Cat3FilterTask' }
+    if cat3_tasks.empty?
+      false
+    else
+      cat3_tasks.first
+    end
+  end
+
+  def task_status(task_type)
+    begin
+      task = tasks.find_by(_type: task_type)
+    rescue
+      return 'incomplete'
+    end
+    task.status
+  end
+
+  def pick_filter_criteria
+    return unless options && options['filters']
+    # select a random patient
+    rand_record = records.sample
+    # iterate over the filters and assign random codes
+    options['filters'].each do |k, v|
+      next if v.count > 0
+      case k
+      when 'races'
+        v << rand_record.race['code']
+      when 'ethnicities'
+        v << rand_record.ethnicity['code']
+      when 'genders'
+        v << rand_record.gender
+      when 'payers'
+        v << rand_record.insurance_providers.first.name
+      when 'providers'
+        options['filters']['providers'] = lookup_provider(rand_record)
+      when 'problems'
+        v << lookup_problem
+      end
+    end
+    save!
+  end
+
+  def lookup_provider(record)
+    provider = Provider.find(record.provider_performances.first['provider_id'])
+    addresses = []
+    provider.addresses.each do |address|
+      addresses << { 'street' => address.street, 'city' => address.city, 'state' => address.state, 'zip' => address.zip,
+                     'country' => address.country }
+    end
+    { 'npis' => [provider.npi], 'tins' => [provider.tin], 'addresses' => addresses }
+  end
+
+  def lookup_problem
```

Comment:
One concern here is that a measure may have multiple diagnosis criteria in a union so it's possible that no records might have a given diagnosis. I think we want to ensure that we always pick a diagnosis that some record actually has. Not sure what the best way to do that is though. Loop & test each possibility?  

Also that raises the question is it possible for a set of records to not have any diagnosis on them when there is a diagnosis criteria? (for example, if the only diagnosis criteria in the measure is unioned with an encounter or something)